Sources/PackageSyntax/ManifestRewriter.swift

```diff
@@ -0,0 +1,822 @@
+            packageDependencies = existingPackageDependencies
+        case .missing:
+            // We didn't find a dependencies section, so insert one.
+            let argListWithDependencies = EmptyArrayArgumentWriter(argumentLabel: "dependencies",
+                                                                   followingArgumentLabels:
+                                                                   "targets",
+                                                                   "swiftLanguageVersions",
+                                                                   "cLanguageStandard",
+                                                                   "cxxLanguageStandard")
+                .visit(initFnExpr.argumentList)
+
+            // Find the inserted section.
+            let packageDependenciesFinder = ArrayExprArgumentFinder(expectedLabel: "dependencies")
+            packageDependenciesFinder.walk(argListWithDependencies)
+            guard case .found(let newPackageDependencies) = packageDependenciesFinder.result else {
+                fatalError("Could not find just inserted dependencies array")
+            }
+            packageDependencies = newPackageDependencies
+        case .incompatibleExpr:
+            diagnosticsEngine.emit(.incompatibleArgument(name: "targets"))
+            throw Diagnostics.fatalError
+        }
+
+        // Add the the package dependency entry.
+        let newManifest = PackageDependencyWriter(
+            name: name,
+            url: url,
+            requirement: requirement,
+            branchAndRevisionConvenienceMethodsSupported: branchAndRevisionConvenienceMethodsSupported
+        ).visit(packageDependencies).root
+
+        self.editedSource = newManifest.as(SourceFileSyntax.self)!
+    }
+
+    /// Add a target dependency.
+    public func addByNameTargetDependency(
+        target: String,
+        dependency: String
+    ) throws {
+        let targetDependencies = try findTargetDependenciesArrayExpr(target: target)
+
+        // Add the target dependency entry.
+        let newManifest = targetDependencies.withAdditionalElementExpr(ExprSyntax(
+            StringLiteralExprSyntax(dependency)
+        )).root
+
+        self.editedSource = newManifest.as(SourceFileSyntax.self)!
+    }
+
+    public func addProductTargetDependency(
+        target: String,
+        product: String,
+        package: TextualPackageReference
+    ) throws {
+        let targetDependencies = try findTargetDependenciesArrayExpr(target: target)
+
+        let dotProductExpr = MemberAccessExprSyntax(base: nil,
+                                                                dot: TokenSyntax.periodToken(),
+                                                                name: TokenSyntax.identifier("product"),
+                                                                declNameArguments: nil)
+
+        let argumentList = TupleExprElementListSyntax([
```

Comment:
This and a few other constructions could be simplified a lot by taking advantage of https://github.com/apple/swift-syntax/pull/848